### PR TITLE
Structured module items

### PR DESCRIPTION
**Pull request by kvark**
This is another small step towards #5 , taken out of #46 
It's based on #62, so only the second commit needs to be reviewed here.
We are now generating a few enums and a few standalone structs for the instructions.

### CODE REVIEW

**Comment by antiagainst** [comment]
This needs a rebase now. :)

**Comment by kvark** [comment]
Rebased now

File: rspirv/sr/autogen_instruction_structs.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+pub struct Capability {
+    capability: spirv::Capability,
+}
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Function {
+    function_control: spirv::FunctionControl,
+    function_type: spirv::Word,
+}
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FunctionParameter {}
```
Comment (antiagainst) [inline rspirv/sr/autogen_instruction_structs.rs]: Do we still want to have separate ops for these function components? I think they should be folded into the function SR itself. I mean, `FunctionEnd` is useful for marking the end of a function in SPIR-V binary, but not useful for SR. Similarly for `Label`, it should be implicit for BasicBlock SR. Letting each pass processing Function or BasicBlock to worry about `FunctionEnd` and `Label` is not helpful.

File: rspirv/sr/items.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    pub capabilities: Vec<op::Capability>,
+    /// All OpExtension instructions.
+    pub extensions: Vec<op::Extension>,
+    /// All OpExtInstImport instructions.
+    pub ext_inst_imports: Vec<op::ExtInstImport>,
+    /// The OpMemoryModel instruction.
+    ///
+    /// Although it is required by the specification to appear exactly once
+    /// per module, we keep it optional here to allow flexibility.
+    pub memory_model: Option<op::MemoryModel>,
```
Comment (antiagainst) [inline rspirv/sr/items.rs]: Do we want to still keep it optional in SR? I think we should mandate it here.

File: rspirv/sr/items.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    pub result: Token<Type>,
+    /// Function parameters.
+    pub parameters: Vec<Token<Type>>,
+    /// Basic blocks in this function.
+    pub basic_blocks: Vec<BasicBlock>,
+}
+
+pub struct Module {
+    /// The module header.
+    pub header: Option<ModuleHeader>,
```
Comment (antiagainst) [inline rspirv/sr/items.rs]: This should be mandated?

File: rspirv/sr/items.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+use spirv;
+
+pub struct EntryPoint {
+    pub execution_model: spirv::ExecutionModel,
+    pub entry_point: Token<Function>,
+    pub name: String,
+    //pub interface: Vec<spirv::Word>,
+}
+
+pub struct BasicBlock {
```
Comment (antiagainst) [inline rspirv/sr/items.rs]: Where is the Vector for instructions?

File: rspirv/sr/autogen_instruction_structs.rs (hunk 1 above)
Comment (kvark) [reply rspirv/sr/autogen_instruction_structs.rs]: I agree, we'll need to set up a list of ops that don't need SR autogen generation, and which we are going to be converting to and from DR manually. We'll do it as a follow-up.

**Comment by kvark** [comment]
Thanks for the review! All the concerns are addressed now, I believe.

**Comment by kvark** [comment]
I think the commit message is descriptive enough? If not, let's talk about this some more and figure out good guidelines.

**Comment by antiagainst** [comment]
I typically prefer to have a concise onliner for describing what the commit is doing. The onliner shouldn't be too long so it renders nicely on GitHub UI (otherwise we will see ... and broken sentences) and also CLI log dump.

Then have some more detailed explanation following after an empty line. I'd prefer to have the detailed explanation to have proper line breaks so each paragraph is not on a single line. (Otherwise when we do `git log`, it's quite annoying to have extra long lines, especially we you use `tmux` or `screen` and are in a small tab, then the message will just be truncated.)

**Comment by antiagainst** [comment]
The content of the message is good (modulo a typo for `module` but not a big thing)! Thanks! :)

**Pull request merged**